src/test/java/hudson/tasks/junit/storage/TestResultStorageTest.java

```diff
@@ -0,0 +1,647 @@
+            assertEquals(1, a.getResult().getPassCount());
+            List<CaseResult> failedTests = a.getFailedTests();
+            assertEquals(2, failedTests.size());
+            assertEquals("Klazz", failedTests.get(0).getClassName());
+            assertEquals("test1", failedTests.get(0).getName());
+            assertEquals("failure", failedTests.get(0).getErrorDetails());
+            assertEquals("another.Klazz", failedTests.get(1).getClassName());
+            assertEquals("test1", failedTests.get(1).getName());
+            assertEquals("another failure", failedTests.get(1).getErrorDetails());
+
+            List<CaseResult> skippedTests = a.getSkippedTests();
+            assertEquals(1, skippedTests.size());
+            assertEquals("other.Klazz", skippedTests.get(0).getClassName());
+            assertEquals("test3", skippedTests.get(0).getName());
+            assertEquals("Not actually run.", skippedTests.get(0).getSkippedMessage());
+
+            List<CaseResult> passedTests = a.getPassedTests();
+            assertEquals(1, passedTests.size());
+            assertEquals("Klazz", passedTests.get(0).getClassName());
+            assertEquals("test2", passedTests.get(0).getName());
+            
+            PackageResult another = a.getResult().byPackage("another");
+            List<CaseResult> packageFailedTests = another.getFailedTests();
+            assertEquals(1, packageFailedTests.size());
+            assertEquals("another.Klazz", packageFailedTests.get(0).getClassName());
+
+            PackageResult other = a.getResult().byPackage("other");
+            List<CaseResult> packageSkippedTests = other.getSkippedTests();
+            assertEquals(1, packageSkippedTests.size());
+            assertEquals("other.Klazz", packageSkippedTests.get(0).getClassName());
+            assertEquals("Not actually run.", packageSkippedTests.get(0).getSkippedMessage());
+
+            PackageResult root = a.getResult().byPackage("(root)");
+            List<CaseResult> rootPassedTests = root.getPassedTests();
+            assertEquals(1, rootPassedTests.size());
+            assertEquals("Klazz", rootPassedTests.get(0).getClassName());
+
+            // TODO test result summary i.e. failure content
+            // TODO getFailedSinceRun, TestResult#getChildren, TestObject#getTestResultAction
+            // TODO more detailed Java queries incl. ClassResult
+            // TODO CaseResult#getRun: In getOwner(), suiteResult.getParent() is null
+            // TODO test healthScaleFactor, descriptions
+            // TODO historyAvailable(History.java:72)
+            // TODO getByPackage UI isn't working in test report
+        }
+    }
+
+    @TestExtension public static class Impl implements TestResultStorage {
+
+        static final String CASE_RESULTS_TABLE = "caseResults";
+
+        static boolean queriesPermitted;
+
+        private final ConnectionSupplier connectionSupplier = new LocalConnectionSupplier();
```

Comment:
I see that this impl has been copied to https://github.com/timja/jenkins-junit-postgresql-plugin/blob/c83e4553ae625e5976f5addaf617402068601266/src/main/java/io/jenkins/plugins/junit/postgresql/PostgresTestResultStorage.java#L39; does it make sense to define a SQL-based implementation in `junit-plugin/src/main/`? Or to define a `junit-sql` plugin with that implementation and add to `test` scope?

Reply:
Possibly tbh I’m not using any Postgres specific features so could just call it junit-sql plugin not Postgres.

It’s very convenient to be able to develop the pluggable storage from inside the junit plugin, possibly once it matures it could be removed from here and then the external reference implementation added as test scope